pkg/controller/datavolume-controller.go

```diff
@@ -0,0 +1,416 @@
+// passed resources of any type other than DataVolume.
+func (c *DataVolumeController) enqueueDataVolume(obj interface{}) {
+	var key string
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	c.workqueue.AddRateLimited(key)
+}
+
+// handleObject will take any resource implementing metav1.Object and attempt
+// to find the DataVolume resource that 'owns' it. It does this by looking at the
+// objects metadata.ownerReferences field for an appropriate OwnerReference.
+// It then enqueues that DataVolume resource to be processed. If the object does not
+// have an appropriate OwnerReference, it will simply be skipped.
+func (c *DataVolumeController) handleObject(obj interface{}) {
+	var object metav1.Object
+	var ok bool
+	if object, ok = obj.(metav1.Object); !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			runtime.HandleError(fmt.Errorf("error decoding object, invalid type"))
+			return
+		}
+		object, ok = tombstone.Obj.(metav1.Object)
+		if !ok {
+			runtime.HandleError(fmt.Errorf("error decoding object tombstone, invalid type"))
+			return
+		}
+		glog.V(4).Infof("Recovered deleted object '%s' from tombstone", object.GetName())
+	}
+	glog.V(4).Infof("Processing object: %s", object.GetName())
+	if ownerRef := metav1.GetControllerOf(object); ownerRef != nil {
+		// If this object is not owned by a DataVolume, we should not do anything more
+		// with it.
+		if ownerRef.Kind != "DataVolume" {
+			return
+		}
+
+		dataVolume, err := c.dataVolumesLister.DataVolumes(object.GetNamespace()).Get(ownerRef.Name)
+		if err != nil {
+			glog.V(4).Infof("ignoring orphaned object '%s' of dataVolume '%s'", object.GetSelfLink(), ownerRef.Name)
+			return
+		}
+
+		c.enqueueDataVolume(dataVolume)
+		return
+	}
+}
+
+// newPersistentVolumeClaim creates a new PVC the DataVolume resource.
+// It also sets the appropriate OwnerReferences on the resource
+// which allows handleObject to discover the DataVolume resource
+// that 'owns' it.
+func newPersistentVolumeClaim(dataVolume *cdiv1.DataVolume) (*corev1.PersistentVolumeClaim, error) {
+	labels := map[string]string{
+		"cdi-controller": dataVolume.Name,
+		"app":            "containerized-data-importer",
+	}
+
+	if dataVolume.Spec.PVC == nil {
+		// TODO remove this requirement and dynamically generate
+		// PVC spec if not present on DataVolume
+		return nil, fmt.Errorf("datavolume.pvc field is required")
+	}
+
+	annotations := make(map[string]string)
+
+	if dataVolume.Spec.Source.HTTP != nil {
+		annotations[AnnEndpoint] = dataVolume.Spec.Source.HTTP.URL
+	} else if dataVolume.Spec.Source.S3 != nil {
+		annotations[AnnEndpoint] = dataVolume.Spec.Source.S3.URL
+		if dataVolume.Spec.Source.S3.SecretRef != "" {
+			annotations[AnnSecret] = dataVolume.Spec.Source.S3.SecretRef
+		}
+	} else {
+		return nil, fmt.Errorf("no source set for datavolume")
```

Comment:
nit: cdi uses the `errors` pkg so that it's easy to get stack traces.

Reply:
fixed :)